Add loss/cost on validation data https://github.com/uma-pi1/kge/issues/2

samuelbroscheit opened the pull request:
As title says, this PR adds loss/cost on validation data to address https://github.com/uma-pi1/kge/issues/2 . Reuse the relevant parts of TrainingJob to run the train job on validation data. This also brings the additional feature that TrainJobs can now run on arbitrary splits; all dependencies on the train split are removed and only determined by the necessary arguments that are propagated down.

rgemulla commented:
Why is this needed? 

We already have the capability to run a training job on arbitrary splits and to select the splits for filtering (in case of negative sampling). E.g., one may simply run a training job using "train.split=valid". 

What may be missing is (1) disable the optimizer, (2) just run one epoch, and (3) disable checkpoints. Is this what this PR is doing?

samuelbroscheit commented:
> We already have the capability to run a training job on arbitrary splits and to select the splits for filtering (in case of negative sampling). E.g., one may simply run a training job using "train.split=valid".
> 
> What may be missing is (1) disable the optimizer, (2) just run one epoch, and (3) disable checkpoints. Is this what this PR is doing?

Yes, there are so many things that have to be disabled (disable optimzer, disable backward, no valid job instantiation, everything in .run() basically, ...) just to run it as a new instance of TrainingJob. I found it to be more sensible to call run_epoch from the valid_job, that has access to its parent_job anyhow. It works very nicely and, of couse, I tested it that it gives the same results as before for all TrainingJobs.

rgemulla commented:
I see & sounds good. Can you very briefly outline the approach/changes? That simplifies the code review.

rgemulla commented:
I do think that we should create a new instance of the training job from the valid job though (so that it can have a different config). That's cleaner and more future-proof than changing code in the training jobs.

rgemulla commented:
This would also allow its use from an arbitrary eval job (not necessarily valid)

samuelbroscheit commented:
> This would also allow its use from an arbitrary eval job (not necessarily valid)

I agree. However, I started with an approach where the TrainingJob was acting depending on a flag in but that made it so ugly  and hard to read with "if"s everywhere. So I ended up with the current solution. I think about if there is another way, because all we care about is run_epoch(). In the current approach it has the signature

```
trace_entry = self.run_epoch(
                split=self.train_split, echo_trace=True, do_backward=True
            )

```

rgemulla commented:
That's fine, but you can still create a new job during valid. Then just call `run_epoch`. The first two arguments are not needed (1st = part of job, 2nd = tresult of run_epoch). I'd rename the third argument to `forward_only` .

samuelbroscheit commented:
Should the training job just produce a seperate trace entry and echo it as well? I think its nicer when we retrieve the trace result and  do

```
            metrics.update(
                avg_loss=train_trace_entry["avg_loss"],
                avg_penalty=train_trace_entry["avg_penalty"],
                avg_penalties=train_trace_entry["avg_penalties"],
                avg_cost=train_trace_entry["avg_cost"],
            )

```

WDYT?

rgemulla commented:
You mean the evaluation job? I'd only copy the avg_loss, the other entries are depending on config settings and do not actually evaluate the model. They can still be found in the trace entry of the subtraining job if needed.

samuelbroscheit commented:
> You mean the evaluation job? I'd only copy the avg_loss, the other entries are depending on config settings and do not actually evaluate the model. They can still be found in the trace entry of the subtraining job if needed.

Ok. Should the subtraining job echo its trace as well?

samuelbroscheit commented:
My current plan is:

create 
`__init_for_train__` and 
`__init_for_eval__`  

which are branched to in `__init__` and 

`run_for_train` and 
`run_for_eval` 

which are branched into from `run()`. The `run_epoch()` trace can be retrieved with a hook.

WDYT

rgemulla commented:
Is this all needed? Why not just stick with `run_epoch` with flag `forward_only`? 

To ease things up, we could additionally create an EvalTrainingLossJob: it creates the TrainingJob, calls run_epoch, and does everything else that's needed.

samuelbroscheit commented:
> Is this all needed? Why not just stick with `run_epoch` with flag `forward_only`?

Ah, did you propose to create the TrainJob but not call run() on it, but only run_epoch? Violates the Job-API a bit, but sure that works. 
 
> To ease things up, we could additionally create an EvalTrainingLossJob: it creates the TrainingJob, calls run_epoch, and does everything else that's needed.

 What would be the advantage of that?

rgemulla commented:
Advantage: (i) that additional job has our Job API and gathers all the "wrapper code" for reuse in other places. E.g., (ii) it could also be run directly from the config (if we wanted that) and (iii) it can be easily used as parts of other jobs (again, if we wanted that). 

For now, key advantage would be (i), not sure if (ii)+(iii) will ever matter.

I am also OK with just using run_epoch for now. It's just that the wrapper job is cleaner, easier to maintain, and probably little additional effort to create.

samuelbroscheit commented:
Alright makes sense and should be easy to do.T


Rainer Gemulla <notifications@github.com> schrieb am Mi., 20. Mai 2020,
14:41:

> Advantage: (i) that additional job has our Job API and gathers all the
> "wrapper code" for reuse in other places. E.g., (ii) it could also be run
> directly from the config (if we wanted that) and (iii) it can be easily
> used as parts of other jobs (again, if we wanted that).
>
> For now, key advantage would be (i), not sure if (ii)+(iii) will ever
> matter.
>
> I am also OK with just using run_epoch for now. It's just that the wrapper
> job is cleaner, easier to maintain, and probably little additional effort
> to create.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/uma-pi1/kge/pull/99#issuecomment-631448686>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFMYSKYFJPUJHW3RKXHWCMDRSPFYPANCNFSM4NDCXCEQ>
> .
>

samuelbroscheit commented:
I wonder if eval.type should not  take a list of eval types, runs all of them, and collects their metrics into one trace.

rgemulla commented:
But this runs multiple eval jobs, right? This could be implemented by having an "aggregate eval" job, which does this. Let's keep this for the future.

samuelbroscheit commented:
Can you review again?

rgemulla commented on kge/job/train.py:
```diff
@@ -538,6 +547,21 @@ def __init__(self, config, dataset, parent_job=None):
+
+        #' for each query type (ordered as in self.query_types), index right after last
+        #' example of that type in the list of all examples
+        self.query_end_index = {}
```
Why was this moved?

samuelbroscheit replied:
Just for consistency. Usually in our codebase we always declare class instance variables in the init(). I can revert if you want.

rgemulla replied:
I see. I suggest to move back for readability. Alternatively, set to `None` here to make clear that this is not correctly initialized.

rgemulla commented on kge/job/train.py:
```diff
@@ -555,21 +579,6 @@ def _prepare(self):
-
         # construct relevant data structures
         self.num_examples = 0
         for query_type in self.query_types:
```
Bit not those parts that fill it? All should stay in prepare

rgemulla commented on kge/job/train.py:
```diff
@@ -756,7 +767,8 @@ def _process_batch(self, batch_index, batch) -> TrainingJob._ProcessBatchResult:
                 forward_time += time.time()
                 backward_time -= time.time()
-                loss_value.backward()
+                if forward_only:
```
if not forward_only (and everywhere else too)

rgemulla commented on kge/job/train.py:
```diff
@@ -812,8 +820,6 @@ def _prepare(self):
             pin_memory=self.config.get("train.pin_memory"),
         )
 
-        self.is_prepared = True
```
Why is anything changed in prepare?

samuelbroscheit replied:
Those were redundant, is already set in run()

rgemulla replied:
Ok

samuelbroscheit commented on kge/job/eval.py:
```diff
@@ -30,6 +33,12 @@ def __init__(self, config, dataset, parent_job, model):
+        train_job_on_eval_split_config.set("train.split", self.eval_split)
+        self.eval_train_loss_job = TrainingJob.create(
+            config=train_job_on_eval_split_config, parent_job=self, dataset=dataset
+        )
```
So instanciate `EvalTrainingLossJob` here? Then `EvalTrainingLossJob.run()` returns the trace and also needs to take a "verbose" argument somewhere. Should we add a `verbose` argument to `Job` which defaults to `True`?

rgemulla replied:
Yes, instantiate the job. 

Adding verbose to job is a good idea, but that requires even more changes (e.g., to apply it consistently) and is better suited to another PR. 

For now, I suggest to use a default-true `verbose` argument only for evaltraininglossjob.

samuelbroscheit commented on kge/job/eval.py:
```diff
@@ -98,6 +111,72 @@ def resume(self, checkpoint_file=None):
+        for f in self.post_epoch_trace_hooks:
+            f(self, trace_entry)
+
+        # if validation metric is not present, try to compute it
```
If EvalTrainingLossJob is run standalone and the avg_loss should be postprocessed and used as a metric then it needs to be here. But as this is a general thing that applies to all eval jobs I made it a post_process hook that is run by all eval jobs, i.e., it didn't belong into only entity ranking as well.

samuelbroscheit replied:
See commit message.

samuelbroscheit commented:
I adressed your reviews. 

I ended up doing a small revision of EvaluationJob: EvaluationJob.run() is implemented now in EvaluationJob which does standard stuff that has to be done for every EvaluationJob and then calls self._run() (suggestions for better name for self.._run() ?).

I shifted EntityRanking specific stuff from EvaluationJob to there.

I added self.verbose to EvaluationJob and use it. 

I made EvalTrainingLossJob an instance variable of EvaluationJob and run it as post_epoch_trace_hook.  This can be made nicer when we create AggregatingEvalJob that runs and aggregates a list of EvaluationJobs. Then the post_epoch_trace_hook can be removed from EvaluationJob and can be replaced by the entry "eval.type: entity_ranking, training_loss" in config-default-yaml. See https://github.com/uma-pi1/kge/issues/102

rgemulla commented on kge/job/train.py:
```diff
@@ -51,21 +52,23 @@ def __init__(
         self.is_prepared = False
         self.model.train()
 
+        if config.get("job.type") == "train":
```
I suggest to disable validation with an optional constructor argument instead of this more hacky approach.

samuelbroscheit replied:
Only as constructor or as a config?

rgemulla replied:
In the config, we have it already. That being said, forget the constructor argument and simply set "valid.every" to 0 in the TrainingLossEvacluatinJob.

rgemulla replied:
Likewise, set train.checkpoint.every to 0.

samuelbroscheit replied:
Hm, I just realized that I was still initializing optimizer (and loss). Optimizer creates new parameters based on model.parameters(). In this context querying job.type makes more sense. See new commit.

rgemulla replied:
You are assuming that the outside eval job does not have job.type=train (it could have). Either set this outside or add a constructor argument "initialize_for_forward_only=False" (preferred since clearer, I think).

samuelbroscheit replied:
> You are assuming that the outside eval job does not have job.type=train (it could have). 

Huh? Isn't it one config == one job? "job.type" is either eval, train, search ..... When I pass trough the eval's config, then its job.type is "eval". I must misunderstand what you are saying here?

> Either set this outside or add a constructor argument "initialize_for_forward_only=False" (preferred since clearer, I think).

Ok can do.

rgemulla replied:
Not if the job is created via EvaluationJob.create(...). An explicit argument makes this clearer

rgemulla commented on kge/job/eval.py:
```diff
@@ -98,6 +166,50 @@ def resume(self, checkpoint_file=None):
         )
 
 
+class EvalTrainingLossJob(EvaluationJob):
```
Why not TrainingLossEvaluationJob for consistency?

rgemulla commented on kge/job/train.py:
```diff
@@ -621,7 +635,7 @@ def collate(batch):
             for example_index in batch:
                 start = 0
                 for query_type_index, query_type in enumerate(self.query_types):
```
drop enumerate if not needed anymore

rgemulla commented on kge/job/train.py:
```diff
@@ -638,7 +652,7 @@ def collate(batch):
             for batch_index, example_index in enumerate(batch):
                 start = 0
                 for query_type_index, query_type in enumerate(self.query_types):
```
likewise

rgemulla commented:
Please also add a CHANGELOG entry before you merge this PR.

rgemulla commented on kge/job/eval.py:
```diff
@@ -175,13 +175,14 @@ def __init__(self, config: Config, dataset: Dataset, parent_job, model):
 
         train_job_on_eval_split_config = config.clone()
         train_job_on_eval_split_config.set("train.split", self.eval_split)
+        train_job_on_eval_split_config.set("negative_sampling.filtering.split", self.config.get("train.split"))
```
Only do this if unset.

rgemulla commented on kge/util/sampler.py:
```diff
@@ -148,7 +148,12 @@ def _filter_and_resample(
+            positives = (
+                index.get(pair).numpy()
+                if pair in index
+                else torch.IntTensor([]).numpy()
```
This tensor should contain the positive and not be empty

rgemulla replied:
Also, why not create a numpy array right away?

rgemulla replied:
Still open.

rgemulla commented on kge/util/sampler.py:
```diff
@@ -277,13 +280,18 @@ def _filter_and_resample_fast(
+            positives_index[pair] = (
+                index.get(pair).numpy()
+                if pair in index
+                else torch.IntTensor([]).numpy()
```
likewise here

rgemulla commented:
Your last changes raise the question of whether all training jobs handle indexes correctly when used in an "eval only" setting. E.g., for KvsAll, what are the labels being used? Training data? Training + eval data?

samuelbroscheit commented:
> Your last changes raise the question of whether all training jobs handle indexes correctly when used in an "eval only" setting. E.g., for KvsAll, what are the labels being used? Training data? Training + eval data?

True, for eval only KvsAll should use Training + eval data.

rgemulla commented:
For the other jobs it's clearer, I guess, but I feel taht loss can turn out to be misleading.

I think we should not make the training_loss a default for the evaluation job (i.e., do not add it as a hook). It may be too confusing for a default since it depends on training/setup so much.

samuelbroscheit commented:
> For the other jobs it's clearer, I guess, but I feel taht loss can turn out to be misleading.
> 
> I think we should not make the training_loss a default for the evaluation job (i.e., do not add it as a hook). It may be too confusing for a default since it depends on training/setup so much.

Actually also for KvsAll its clear, its the same as for all losses. It should only be the labels from eval data, i.e. for unseen labels. 

The only particular thing with NS is, that the current implementation does the filtering of positive triples based on what it is given as `train.split`. But actually that is always the known KG, i.e., the train split.

rgemulla commented:
Why is it clear for KvsAll? If done like you state, the losses are not comparable between train and valid (valid loss will be pretty much like 1vsAll because the data probably looks like this). Nevertheless this may be the best option. Another sensible choice may be the prefixes/suffixes from eval, but the labels from train+eval.

rgemulla commented:
For negative sampling, the filtering split (or splits?) can already be specified, I think.

samuelbroscheit commented:
> For negative sampling, the filtering split (or splits?) can already be specified, I think.

Yes but at the moment it is automagically the split defined as `train.split`. However, `train.split` is set to "valid" when TrainingJob is run as evaluation. https://github.com/uma-pi1/kge/blob/f2bcbdb47776e6a3b73d07cca28603bbce723f88/kge/util/sampler.py#L33-L34

rgemulla commented:
> > For negative sampling, the filtering split (or splits?) can already be specified, I think.
> 
> Yes but at the moment it is automagically the split defined as `train.split` if not set otherwise.
So which one to pick here? The most natural choice would be train+valid, but that does not seem to be supported.

samuelbroscheit commented:
> Why is it clear for KvsAll? If done like you state, the losses are not comparable between train and valid (valid loss will be pretty much like 1vsAll because the data probably looks like this). Nevertheless this may be the best option. Another sensible choice may be the prefixes/suffixes from eval, but the labels from train+eval.

When we were talking about train+eval, this is what I meant, prefixes from valid, but labels from both. I understand what you mean that the loss on train and the loss on valid are not comparable in that case. With "it is clear" I meant that the only-on-valid KvsAll loss would still work to measure overfitting. Would be nicer if it would be a comparable loss, though.

samuelbroscheit commented:
> > > For negative sampling, the filtering split (or splits?) can already be specified, I think.
> > 
> > 
> > Yes but at the moment it is automagically the split defined as `train.split` if not set otherwise.
> 
> So which one to pick here? The most natural choice would be train+valid, but that does not seem to be supported.

So should we?

rgemulla commented:
If we allow filtering with train+valid, we'd need to change the interface. Not sure if we want to go there; it significantly complicates the code if we allow for multiple splits during training.

An alternative may be to create new splits by merging existing splits. That's wasting memory, but at least it's less intrusive. The easiest way to do this is (1) extend the dataset.yaml with support for multple files per split, (2) add a train_valid and a train_valid_test split with the appropriate files. That should be a separate PR. For now, why not simply use the valid split and issue a warning or an error when KvsAll is used like that?

rgemulla commented:
Also, using KvsAll with a different "query split" and "label split" is highly confusing. 

Option 1: TrainingLossEvaluationJob solely uses the validation split and ignores all other splits (also for negative sampling etc). This way, it's clear what it does and we can issue a warning saying this. How useful such a loss is, is another question.

Option 2: we do not add this method of evaluation. For comparabiliy with training data, we can proceed the other way around: run the standard eval job on a sample of the training data (this sampling can be done externally to create a new train_sample split).

rgemulla commented:
I tend to option 2.

samuelbroscheit commented:
I'm for option 3: do it and do it correctly. 

Now we're almost there and understand the implications. Would be a waste, now that we scrutinized it. I also like that this is tightly coupled to the filtering of entity_ranking. This would mean that EvaluationJob and TrainingJob both have a query_split and a label_splits attribute. label_splits would replace filter_splits in EvaluationJob. It stays an automatically set option and is not a configurable option. We can also call it "known answers". 

For the TrainingJob when not in eval mode -> nothing changes. query split == label split. 

In eval mode:
- For 1vsAll in eval mode: label_split == query_split. 

- For NS in eval mode: label_split only comes into play when positives are filtered -> label_split = train split + eval split(s)

- KvsAll  in eval mode:  label_split are the labels -> label_split = train split + eval split(s)
For KvsAll we do not need extra dataset files. Its easy to merge the dictionaries and remove the keys that are not in train_split.

rgemulla commented:
That would also be a suitable option. The main drawback is that it's more work. 

I suggest a slightly revised approach:

For 1vsAll: do nothing
For NS: add support for multiple filter splits (not just one, as we have now)
For KvsAll: add a label_splits option (under key KvsAll) along the lines you discussed

For NS: rename negative_sampling.filtering.split to negative_sampling.filtering.splits. Automatically add train.split, but additionally use the splits mentioned here (along the lines we already do it for eval.filter_splits).

For KvsAll: You are right in that the label_splits do not need to be merged apriori (which would mean that we need to copy all data). This can be done in the respective collate() function during data loading. If label_splits is unset, train.split should be used. If set, these splits + train.split should be used (again, along the lines we already do it for eval.filter_splits).

samuelbroscheit commented:
https://github.com/uma-pi1/kge/pull/99/commits/083b27cd977175d7aa30e7467adf5f1c36055eea contains all the relevant changes

samuelbroscheit commented:
Can you review https://github.com/uma-pi1/kge/commit/66a9bc17d9232a8172cb151df4d4dbfed9b638f3 or is there anything else?

rgemulla commented:
I cannot review this PR as such. Instead of rebasing it on master, it contains tons of master commits which make it hard to review. Can you rebase on master so that just the changes are here?

samuelbroscheit commented:
Now?

rgemulla commented:
Even worse, you destroyed it. Do not touch, I force push in a minute.

rgemulla commented on kge/config-default.yaml:
```diff
@@ -235,6 +235,10 @@ KvsAll:
 
+  # Dataset splits from which the labels for a query are taken from. Default: If
+  # nothing is specified, then the train split is used.
+  label_splits: []
```
Semantics should be: additional splits to train split.

rgemulla commented on kge/config-default.yaml:
```diff
@@ -264,7 +268,7 @@ negative_sampling:
     o: False       # as above
 
-    split: ''      # split containing the positives; default is train.split
+    splits: []     # splits containing the positives; default is train.split
```
Likewise here.

rgemulla commented on kge/config-default.yaml:
```diff
@@ -327,7 +331,8 @@ eval:
 
-  # Type of evaluation (entity_ranking only at the moment)
+  # Type of evaluation (entity_ranking or training_loss). Currently, 
+  # entity_ranking runs training_loss as well.
```
I vote for disabling the automatic running of training_loss. Makes evaluation too costly.

samuelbroscheit replied:
Validation and test data are a lot smaller (1/10 is the worst valid/train ratio in our current datasets).

rgemulla commented on kge/indexing.py:
```diff
@@ -222,13 +223,31 @@ def _invert_ids(dataset, obj: str):
+    index_name = f"{split_combi_str}_{key}_to_{value}"
+    indexes = [dataset.index(f"{_split}_{key}_to_{value}") for _split in split]
+    dataset._indexes[index_name] = merge_dicts_of_1dim_torch_tensors(indexes)
+    return dataset._indexes[index_name]
```
Instead of merging indexes (very expensive memory wise), the code should support using multiple indexes.

samuelbroscheit replied:
For NS I see that it works, for KvsAll I have to merge them. But granted, not neccessarily as an index.

rgemulla replied:
What I meant here: the correspoding training jobs should support this. Alternatively, one may write a wrapper index that takes muliple indexes and merges there results on demand.

rgemulla replied:
For KvsAll, you can also just add the labels of each label split sequentially in the collate function.

rgemulla replied:
We should not allow using label_splits other than for forward only. Otherwise they leak valid/test data. Likewise for negative sampling, I guess.

rgemulla commented on kge/indexing.py:
```diff
@@ -222,13 +223,31 @@ def _invert_ids(dataset, obj: str):
+            index_name = f"{split_combi_str}_{key}_to_{value}"
+            dataset.index_functions[index_name] = IndexWrapper(
+                merge_KvsAll_indexes, split=split_combi, key=key
+            )
```
This makes it even worse.

samuelbroscheit replied:
Why? They are not invoked. 

As a sidenote: This was neccasary because there is currently no way to determine the cross-validation semantics of a dataset file (train,valid,test). We should (in another PR) add a xvaltype attribute to the dataset files.

rgemulla replied:
There are also not needed. Let's keep the code clean and not add "orphans"

rgemulla replied:
Adding more information to the dataset configuration is a good idea (if useful for something).

rgemulla commented on kge/job/eval.py:
```diff
@@ -31,9 +25,11 @@ def __init__(self, config, dataset, parent_job, model):
-        self.filter_with_test = config.get("eval.filter_with_test")
         self.epoch = -1
 
+        self.verbose = True
```
Drop this. NOt needed anymore.

rgemulla commented on kge/job/eval.py:
```diff
@@ -31,9 +25,11 @@ def __init__(self, config, dataset, parent_job, model):
         self.filter_splits = self.config.get("eval.filter_splits")
         if self.eval_split not in self.filter_splits:
             self.filter_splits.append(self.eval_split)
-        self.filter_with_test = config.get("eval.filter_with_test")
```
accidental delete? where was this moved?

samuelbroscheit replied:
entity_ranking

rgemulla commented on kge/job/eval.py:
```diff
@@ -64,6 +60,22 @@ def __init__(self, config, dataset, parent_job, model):
+                lambda job, trace: trace.update(
+                    avg_loss=self.eval_train_loss_job.run()["avg_loss"]
+                )
+            )
```
drop this.

rgemulla commented on kge/job/eval.py:
```diff
@@ -132,8 +200,65 @@ def create_from(
+
+    def __init__(self, config: Config, dataset: Dataset, parent_job, model):
+        super().__init__(config, dataset, parent_job, model)
+        self.is_prepared = True
```
Don't do this. Simply use an empty _prepare method. Perhaps you can even let it emtpy in the supervlass instead of using NotImplementedError

rgemulla commented on kge/misc.py:
```diff
@@ -15,6 +18,33 @@ def is_number(s, number_type):
         return False
 
 
+def merge_dicts_of_1dim_torch_tensors(keys_from_dols, values_from_dols=None):
```
these methods can be dropped with the changes suggested above

samuelbroscheit replied:
I still need this method for KvsAll, right?

rgemulla commented on kge/util/sampler.py:
```diff
@@ -29,9 +29,9 @@ def __init__(self, config: Config, configuration_key: str, dataset: Dataset):
-            self.filtering_split = config.get("train.split")
+        self.filtering_splits: List[str] = config.get("negative_sampling.filtering.splits")
+        if len(self.filtering_splits) == 0:
+            self.filtering_splits.append(config.get("train.split"))
```
always append when not there. also, add handling of depracated configs to config.py

rgemulla commented on kge/util/sampler.py:
```diff
@@ -43,7 +43,10 @@ def __init__(self, config: Config, configuration_key: str, dataset: Dataset):
+                for filtering_split in self.filtering_splits:
+                    dataset.index(f"{filtering_split}_{pair}_to_{slot_str}")
+                filtering_splits_str = '_'.join(sorted(self.filtering_splits))
+                dataset.index(f"{filtering_splits_str}_{pair}_to_{slot_str}")
```
drop

rgemulla commented on kge/util/sampler.py:
```diff
@@ -279,13 +287,18 @@ def _filter_and_resample_fast(
-            positives_index[pair] = index.get(pair).numpy()
+            positives_index[pair] = (
+                index.get(pair).numpy()
+                if pair in index
```
instead, check here for whether it's in one of the indexes etc

samuelbroscheit commented:
> I rebased (hopefully correctly) and added some review comments.

No its a bit broken, but its fixable.

rgemulla commented:
Yes, I did not know your new code well enough at places. I think all the echo_true and verbose things need to go out, but you'll be much quicker with this.
